spec/aptible/cli/agent_spec.rb

```diff
@@ -155,17 +160,76 @@
           expect(Aptible::Auth::Token).to receive(:create)
             .with(email: email, password: password, expires_in: 1.week.seconds)
             .once
-            .and_raise(OAuth2::Error, 'otp_token_required')
+            .and_raise(make_oauth2_error('otp_token_required'))
 
           expect(Aptible::Auth::Token).to receive(:create)
             .with(email: email, password: password, otp_token: token,
                   expires_in: 12.hours.seconds)
             .once
-            .and_raise(OAuth2::Error, 'foo')
+            .and_raise(make_oauth2_error('foo'))
 
           expect { subject.login }.to raise_error(/Could not authenticate/)
         end
       end
+
+      context 'with U2F' do
```

Comment:
> Do we run any integration tests for this against a test or staging API server? That may be the better option here get a true sense of if this is functional.

We do include the CLI in https://github.com/aptible/aptible-integration, but I don't we'd be able to actively test hardware based authentication without a human and authentication device?

Reply:
Yeah, I don't have perfect suggestion for this. Was more just musing in that comment. I was sort of thinking of https://github.com/github/SoftU2F as I way get around the human bit, but I doubt it'd be useful here. 

The second para about thinking about how to test out scenarios with a "mocked" `u2f-host` was the more pertinent thought. The logic around spawning and waiting on the subprocesses and using their exit codes and return values _feels_ like a place we'd want to specify and test behavior. It looks subtle to me and may be a place to increase confidence.

I don't think that's a blocker here at all. The risk is rather low on it and the effort to improve testing probably is not worth it right now. Just wanted to note it in case anyone had a brighter idea than me 😄